src/control/connector.rs

```diff
@@ -252,3 +258,53 @@ impl From<State> for u32 {
         }
     }
 }
+
+/// Subpixel order of the connected sink
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub enum SubPixel {
+    /// Unknown geometry
+    Unknown,
+    /// Horizontal RGB
+    HorizontalRgb,
+    /// Horizontal BGR
+    HorizontalBgr,
+    /// Vertical RGB
+    VerticalRgb,
+    /// Vertical BGR
+    VerticalBgr,
+    /// No geometry
+    None,
+}
+
+impl From<u32> for SubPixel {
+    fn from(n: u32) -> Self {
+        // These values are not defined in drm_mode.h
+        // They were copied from linux's drm_connector.h
+        // Don't mistake those for ones used in xf86DrmMode.h (libdrm offsets those by 1)
+        match n {
+            0 => Self::Unknown,
+            1 => Self::HorizontalRgb,
+            2 => Self::HorizontalBgr,
+            3 => Self::VerticalRgb,
+            4 => Self::VerticalBgr,
+            5 => Self::None,
+            _ => Self::Unknown,
```

Comment:
Feels like this should be an `Err` via a `try_from()`, so that if any new geometries were ever added they aren't treated the same as the known value `0`?

Reply:
If that's the case, we should probably do the same as #143 and make it `#[non_exhaustive]` to allow the ability for new formats (and since it's a new type, no breaking change is made :+1: )

Also, if we're getting these values directly from the kernel, then it might be simpler to make the enum `#[repr(u32)]` and set the values directly in the definition. We'd still need to check the value when we convert to the type, but would provide some guarantees if we try to use it in some other way.

As far as returning `Self::Unknown` vs `Err` (in the case of a new format being added to the kernel), I don't think the `Err` is useful downstream unless we also include the original, unknown value. And if their code checks that value and tries to act on it in some way, then adding the new format would result in being a breaking change for the downstream user, as they would no longer be receiving the `Err` at all to act on. So I'd recommend sticking with `Self::Unknown` unless there's another reason.

Reply:
> Also, if we're getting these values directly from the kernel, then it might be simpler to make the enum `#[repr(u32)]` and set the values directly in the definition.

Are there more enums already in this cratre that could/should get the same treatment? Perhaps `enum Interface` that was updated in #143?

Reply:
@Slabity I wasn't actually suggesting to use `Err(6)` directly (the `UnknownValue` could be a ZST struct or enum value without contents), and the user can simply map the original value in that case since `SubPixel` is `Copy`. Thanks to `non_exhaustive` their code should also catch new unhandled `enum` values via however way they wish to handle it.

After all this would be a kernel API change/extension, and the only thing drm-rs can do about this is communicating it as transparently to the user as possible, without possibly giving the ambiguity of having `Err(6)` at some point and a `HorizontalRGBY` later.

Reply:
Made the whole thing private, subpixel() getter will still return unknown as is did, if we want to change the way we handle enums I would do that in all enums at the same time.

In addition to that, as stated in the second thread, I believe using values different than xf86DrmMode here would only confuse people, so keeping it private may simply be the best choice.

Reply:
Still disagree about mapping `0` and `> 5` to `Unknown`, making the user unable whatsoever to intercept new values, unlike an `UnknownValue(u32)`, but alas.

Reply:
I agree, that the Unknown variant could have an integer value, but then all of them should sooo...

Given it is not_exhaustive, I feel like this is good to merge until we made a definite decision on this type of enums for the whole project? @Slabity any objections?

I tend to agree with @Slabity here in the general case though. I think we should avoid breakage and are quick to update the crate anyway, if anyone wants to add new variants, so keeping this up-to-date doesn't seem like an issue to me.

Reply:
But wouldn't that mean a new kernel (returning a new value) could potentially break existing binaries? That doesn't sound too great to me.

Reply:
Correct.

Do we expect to extend this enum value often? If not, perhaps it is safer to drop `non_exhaustive` and have an explicit `UnknownValue(u32)` (for all enums across this crate if it makes things consistent). That removes ambiguity, as when we add the theoretical `HorizontalRGBY = 6` in a breaking release downstream code is required to extend their `match` statement rather than having this end up in the catch-all arm that they are forced to have?

Reply:
> Do we expect to extend this enum value often?

No we do not, which is why I think updating drm-rs once new values actually get added is reasonable from a maintenance perspective.
Which means we could have a stable api and stable binaries at the cost of having to extend this enum, if it ever gets updated.

Which seems to be the better trait off to me than either having binaries panic on newer kernels or having matches on `Unknown`-variants with values break in the future.

> If not, perhaps it is safer to drop non_exhaustive and have an explicit UnknownValue(u32) (for all enums across this crate if it makes things consistent).

That would also be an option that satisfies these constraints. The only immediate benefit I see in this approach is, that it doesn't force users to upgrade to newer drm-rs releases to handle new values. Which might be nice to handle compatibilities with other crates or rust versions, though we don't frequently bump the MSRV and most accompanying crates (like gbm.rs) have a release almost immediately following drm-rs.

So the solutions are either:
- Opaque enum (raw values inaccessible including) and non-exhaustive.
- Visible unknown values and exhaustive.

The latter is more flexible, but requires more frequent major version bumps by semver. Which is why I slightly prefer the former, but either is fine to me.

Reply:
@Slabity would you mind making a decision here? I think all pros and cons have been clearly layed out.

Reply:
@Drakulix yeah, thanks for laying out the two options more explicitly and concisely.

I think I prefer the first option with `non_exhaustive` and a non-opaque unknown variant as well, as long as it's different from `Unknown = 0`. That both enables `drm-rs` to do frequent _minor_ bumps, as well as force crate authors to update often if they wish to handle this case by not giving them access to the value.

It has been a while since I used `non_exhaustive` however: is there already a lint that is able to tell you if there are new variants in the `enum` that a user does not match on, because they are forced to have a catch-all `_ => {...}` arm? To give more visibility into such changes in updates.

Reply:
> It has been a while since I used non_exhaustive however: is there already a lint that is able to tell you if there are new variants in the enum that a user does not match on, because they are forced to have a catch-all _ => {...} arm? To give more visibility into such changes in updates.

If there isn't that is something we should open a `clippy`-issue on.

Reply:
> I think I prefer the first option with non_exhaustive and a non-opaque unknown variant as well, as long as it's different from Unknown = 0

If the unknown variant is non-opaque we run into the same problem again. Either everything is opaque and Unknown can't be differentiated from a legitimately unknown variant + non_exhaustive **or** nothing is opaque and it's also exhaustive.

Reply:
@Drakulix whoops I meant to say "opaque", but indeed we'd still have to come up with a plausible discriminant for it that doesn't overlap with a possible kernel value in the future - and that's exactly impossible when the kernel value is `u32` and the `enum` is `repr(u32)` too.

Reply:
Right, which is why the enum can't be `repr(u32)`. We can add a new variant `Invalid`, if you want to explicitly differentiate between "unknown value" and "unknown as a value".

Reply:
In one of the crates that I maintain where we use `num_enum`, we're currently debating a similar issue. There's now a `num_enum(default)` for such an `Invalid` variant, and a `num_enum(catch_all)` for an `Invalid(u32)` variant. This does blow up the `enum` as it generates every variant _with discriminant_ (and uses that to cast everything _but_ the `Invalid` variant), meaning extra space is needed to store the `u32` for `Invariant(u32)`.

I'm still on the fence there which way to go :sweat_smile:

Reply:
As I said I don't want to expose the raw value, if we go `non_exhaustive`, which I would prefer as an api over an exhaustive variant. I am thus not sure, that an opaque `Invalid` variant has any advantage over just returning `Unknown`, because frankly that is how you should treat the value anyway in that case.

Reply:
Sorry guys, I was away camping for a bit and didn't see the pings.

The major issue I see with using a variant that returns the raw value is that once we *do* add the value properly, it will break everyone's code that uses that variant regardless. If a user matches on something like `Invalid(11)` (or whatever it would be called) and acts on it in some way, then we add `11 => Self::QuantumSuperpositionRGBXYZ` to the enum, then that code path *they* added will break. And this can potentially break *every time we add one*. There's a few different things we can do to get around that, but honestly if someone needs that, then it likely wouldn't be too big of a deal to have them file an issue or open a PR and add a line to fix it.

On the other hand, I don't really want to have people wait for an unknown amount of time before we update the crate to match new values that get implemented upstream. It could get slightly annoying to have these values get implemented from kernel -> `x86drm*` bindings -> `drm-rs` bindings. It won't happen often in this case, but there's other cases where this might potentially happen much more often.

Instead, maybe it would be feasible to try and get these automated? There's not really a lot of human complexity to just add a new value to an enum. It would take some effort at first, but would allow people to update it on their own *and* we'd be able to leverage CI to make maintaining these sort of changes way easier.

Reply:
No worries.

Again, I agree that having an opaque value prevents ambiguity, by _not allowing a user to match on `Invalid(11)`_.

> but honestly if someone needs that, then it likely wouldn't be too big of a deal to have them file an issue or open a PR and add a line to fix it.

Again^2, I also think that mapping **both** `0 => Unknown` **and** `_ => Unknown` actively stalls end-users in reporting such issues. They now have no way (without casting or inspecting structures?) to tell apart whether:

1. The kernel returned `0`;
2. The kernel returned **non-`0`** and I need to ask `drm-rs` to map the new value.

That's why I suggested to cook up a different variant (e.g. `Invalid`) for the `_` case.

Reply:
Ah, sorry, I misinterpreted. I understand now. By adding `Invalid` as its own value (different from the kernel's `Unknown`) we it can be caught downstream and get people to report it more quickly.

That makes a lot of sense, I think that's probably a good idea so long as we document it as such to tell downstream users something like, "Hey, this means *we* don't know what it is yet, not that your system doesn't"

I agree with adding that. Though maybe not calling it `Invalid` as that might make people think the kernel is sending an actual invalid value. Maybe `Unimplemented`?

Reply:
Yeah I wasn't too happy about the `Invalid` brought up above; I do like the suggested `Unimplemented` together with an appropriate doc-comment.

Just in case (the outcome of) this long trail of discussion should probably end up in some doc or guideline as it transcends this PR and the simplicity of Subpixel in general. I'm taking these points ij other crates into account as well.

Reply:
Sounds good. I won't bikeshed too much on what to call it, but either `Unimplemented` or `NotImplemented` would be fine by me. The latter *might* make it be less associated with the `unimplemented!` macro but honestly I'm not too picky.